File: backend/voting-app/src/main/java/com/tu/votingapp/entities/elections/VoteEntity.java
package com.tu.votingapp.entities.elections;

import com.tu.votingapp.entities.UserEntity;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.sql.Date;

@Entity
@Table(name = "votes", uniqueConstraints = {
        @UniqueConstraint(columnNames = {"user_id", "election_id"})
})
@Data
@AllArgsConstructor
@NoArgsConstructor
public class VoteEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false) // A vote MUST belong to a user
    @JoinColumn(name = "user_id", nullable = false) // Foreign key column
    private UserEntity user; // Changed from Long userId

    @ManyToOne(fetch = FetchType.LAZY, optional = false) // A vote MUST belong to an election
    @JoinColumn(name = "election_id", nullable = false)
    private ElectionEntity election;

    // --- Added Fields for Vote Choice ---

    @ManyToOne(fetch = FetchType.LAZY, optional = true) // Vote might be for a party only, so candidate is optional
    @JoinColumn(name = "candidate_id", nullable = true) // Allow NULL if voting only for party
    private CandidateEntity candidate; // The specific candidate voted for (if applicable)

    @ManyToOne(fetch = FetchType.LAZY, optional = true) // Vote might be for an independent candidate, so party is optional
    @JoinColumn(name = "party_id", nullable = true) // Allow NULL if voting for independent candidate
    private PartyEntity party; // The party voted for (either directly or the party of the candidate)


    @Column(name = "vote_timestamp")
    private Date voteTimestamp;

    public VoteEntity(Long id) {
        this.id = id;
    }

}

File: frontend/android/app/src/main/java/com/example/android/entity/election/VoteEntity.kt
package com.example.android.entity.election

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey
import com.example.android.db.orm.Column
import com.example.android.db.orm.Table


@Entity(
    tableName = "votes",
    foreignKeys = [
        ForeignKey(
            entity = ElectionEntity::class,
            parentColumns = ["id"],
            childColumns = ["election_id"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = CandidateEntity::class,
            parentColumns = ["id"],
            childColumns = ["candidate_id"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = PartyEntity::class,
            parentColumns = ["id"],
            childColumns = ["party_id"],
            onDelete = ForeignKey.CASCADE
        )
    ]
)
data class VoteEntity(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    val id: Long = 0,

    @ColumnInfo(name = "user_id")
    val userId: Long,

    @ColumnInfo(name = "election_id")
    val electionId: Long,

    @ColumnInfo(name = "candidate_id")
    val candidateId: Long? = null,

    @ColumnInfo(name = "party_id")
    val partyId: Long? = null,

    @ColumnInfo(name = "vote_timestamp")
    val voteTimestamp: String // Store as ISO 8601 string, e.g. "2025-04-19T16:00:00"
)